engine/table/src/main/java/io/deephaven/engine/table/impl/locations/TableDataService.java

```diff
@@ -20,6 +20,26 @@ public interface TableDataService {
     @NotNull
     TableLocationProvider getTableLocationProvider(@NotNull TableKey tableKey);
 
+    /**
+     * Request the single raw {@link TableLocationProvider} from this service that provides the {@link TableLocation}
+     * for {@code tableKey} and {@code tableLocationKey}. A raw {@link TableLocationProvider} does not compose multiple
+     * {@link TableLocationProvider TableLocationProviders} or delegate to other implementations.
+     *
+     * @param tableKey The {@link TableKey} to lookup
+     * @param tableLocationKey The {@link TableLocationKey} to lookup
+     * @return The raw {@link TableLocationProvider} that provides the {@link TableLocation} for {@code tableKey} and
+     *         {@code tableLocationKey}, or {@code null} if there is none
+     * @implSpec Non-raw {@link TableDataService TableDataServices} must implement this method to provide a single raw
+     *           {@link TableLocationProvider} or throw a {@link TableDataException} if there is ambiguity.
+     * @throws TableDataException If the {@link TableLocation} is provided by more than one
+     *         {@link TableLocationProvider}
+     */
+    @Nullable
+    default TableLocationProvider getRawTableLocationProvider(@NotNull final TableKey tableKey,
+            @NotNull final TableLocationKey tableLocationKey) {
+        return getTableLocationProvider(tableKey);
```

Comment:
The implementations I can find of getTableLocationProvider are "putIfAbsent", so if you manage to call getRawTableLocationProvider on an eligible (e.g. filtered) TDS, it will return a TLP even if it has no locations for the key.